// @astrojs/telemetry/dist/events/session.js
import { createRequire } from "node:module";
const require2 = createRequire(import.meta.url);
const EVENT_SESSION = "ASTRO_CLI_SESSION_STARTED";
function getViteVersion() {
  try {
    const { version } = require2("vite/package.json");
    return version;
  } catch (e) {
  }
  return void 0;
}
const multiLevelKeys = /* @__PURE__ */ new Set([
  "build",
  "markdown",
  "markdown.shikiConfig",
  "server",
  "vite",
  "vite.resolve",
  "vite.css",
  "vite.json",
  "vite.server",
  "vite.server.fs",
  "vite.build",
  "vite.preview",
  "vite.optimizeDeps",
  "vite.ssr",
  "vite.worker"
]);
function configKeys(obj, parentKey) {
  if (!obj) {
    return [];
  }
  return Object.entries(obj).map(([key, value]) => {
    if (typeof value === "object" && !Array.isArray(value)) {
      const localKey = parentKey ? parentKey + "." + key : key;
      if (multiLevelKeys.has(localKey)) {
        let keys = configKeys(value, localKey).map((subkey) => key + "." + subkey);
        keys.unshift(key);
        return keys;
      }
    }
    return key;
  }).flat(1);
}
function eventCliSession(event, userConfig, flags) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j;
  const integrations = ((_b = (_a = userConfig == null ? void 0 : userConfig.integrations) == null ? void 0 : _a.filter) == null ? void 0 : _b.call(_a, Boolean)) ?? [];
  const configValues = userConfig ? {
    markdownPlugins: [
      ((_c = userConfig == null ? void 0 : userConfig.markdown) == null ? void 0 : _c.remarkPlugins) ?? [],
      ((_d = userConfig == null ? void 0 : userConfig.markdown) == null ? void 0 : _d.rehypePlugins) ?? []
    ].flat(1),
    adapter: ((_e = userConfig == null ? void 0 : userConfig.adapter) == null ? void 0 : _e.name) ?? null,
    integrations: ((_f = integrations == null ? void 0 : integrations.map) == null ? void 0 : _f.call(integrations, (i) => i == null ? void 0 : i.name)) ?? [],
    trailingSlash: userConfig == null ? void 0 : userConfig.trailingSlash,
    build: (userConfig == null ? void 0 : userConfig.build) ? {
      format: (_g = userConfig == null ? void 0 : userConfig.build) == null ? void 0 : _g.format
    } : void 0,
    markdown: (userConfig == null ? void 0 : userConfig.markdown) ? {
      mode: (_h = userConfig == null ? void 0 : userConfig.markdown) == null ? void 0 : _h.mode,
      syntaxHighlight: (_i = userConfig.markdown) == null ? void 0 : _i.syntaxHighlight
    } : void 0
  } : void 0;
  const cliFlags = flags ? Object.keys(flags).filter((name) => name != "_") : void 0;
  const payload = {
    cliCommand: event.cliCommand,
    astroVersion: event.astroVersion,
    viteVersion: getViteVersion(),
    nodeVersion: process.version.replace(/^v?/, ""),
    configKeys: userConfig ? configKeys(userConfig, "") : void 0,
    config: configValues,
    flags: cliFlags,
    optionalIntegrations: ((_j = userConfig == null ? void 0 : userConfig.integrations) == null ? void 0 : _j.length) - (integrations == null ? void 0 : integrations.length)
  };
  return [{ eventName: EVENT_SESSION, payload }];
}
export {
  eventCliSession
};
